sync-core/src/main/java/com/cloudant/sync/datastore/BasicDatastore.java

```diff
@@ -1067,90 +1083,115 @@ public Object call(SQLDatabase db) throws Exception{
-                                    for (String[] key : preparedAttachments.keySet()) {
-                                        String id = key[0];
-                                        String rev = key[1];
+                        try {
+                            if (preparedAttachments != null) {
+                                for (String[] key : preparedAttachments.keySet()) {
+                                    String id = key[0];
+                                    String rev = key[1];
+                                    try {
                                         BasicDocumentRevision doc = getDocumentInQueue(db, id, rev);
-                                        if(doc != null) {
-                                            for (PreparedAttachment att : preparedAttachments.get(key)) {
+                                        if (doc != null) {
+                                            for (PreparedAttachment att : preparedAttachments.get
+                                                    (key)) {
                                                 attachmentManager.addAttachment(db, att, doc);
                                             }
                                         }
+                                    } catch (DocumentNotFoundException e){
+                                        continue;
                                     }
                                 }
-                            } catch(Exception e){
-                                logger.log(Level.SEVERE, "There was a problem adding an " +
-                                         "attachment to " +
-                                        "the datastore, terminating replication", e);
-                                throw e;
                             }
-
+                        } catch (Exception e) {
+                            logger.log(Level.SEVERE, "There was a problem adding an " +
+                                    "attachment to " +
+                                    "the datastore, terminating replication", e);
+                            throw e;
                         }
-                        if (ok) {
-                            logger.log(Level.FINER, "Inserted revision: %s", rev);
-                            if (documentCreated != null) {
-                                eventBus.post(documentCreated);
-                            } else if (documentUpdated != null) {
-                                eventBus.post(documentUpdated);
-                            }
+
+                    }
+                    if (ok) {
+                        logger.log(Level.FINER, "Inserted revision: %s", rev);
+                        if (documentCreated != null) {
+                            eventBus.post(documentCreated);
+                        } else if (documentUpdated != null) { //TODO this event *should* not be
+                        // fired on the db queue.
```

Comment:
Is there a bug somewhere to remind us to fix this? Why shouldn't it be called on the queue? Why's it fine for created notifications?